akka-stream/src/main/scala/akka/stream/impl/JsonObjectParser.scala

```diff
@@ -36,6 +37,469 @@ import scala.annotation.switch
+      override def toString: String = "InString"
+    }
+    private object InOuterString extends InStringBase with CompleteObjectOnExitAction {
+      override def toString: String = "InOuterString"
+    }
+
+    private abstract class InNakedBase extends LeafParserState with HasExitAction {
+      final def apply(nextState: ParserState, pp: JsonObjectParser): ParserState = {
+        pp.stateAfterNakedValue = nextState
+        pp.enterState(this)
+        this
+      }
+
+      private def finish(input: Byte, pp: JsonObjectParser): Unit = {
+        val nextState = pp.stateAfterNakedValue
+        pp.stateAfterNakedValue = UnknownState
+
+        pp.enterState(nextState)
+        exitAction(pp)
+        nextState.proceed(input, pp)
+      }
+
+      @tailrec
+      final override def proceed(input: Byte, pp: JsonObjectParser): Unit = input match {
+        case Comma | SquareBraceEnd | CurlyBraceEnd | LineBreak ⇒
+          finish(input, pp)
+
+        case _ if isWhitespace(input) ⇒
+          finish(input, pp)
+
+        case _ if !isWhitespace(input) ⇒
+          pp.take()
+          if (pp.keepSeeking) {
+            proceed(pp.nextInput, pp)
+          }
+      }
+    }
+
+    private object InNaked extends InNakedBase with HasEmptyExitAction {
+      override def toString: String = "InNaked"
+    }
+    private object InOuterNaked extends InNakedBase with CompleteObjectOnExitAction {
+      override def toString: String = "InOuterNaked"
+    }
+
+    private abstract class InContainerBase(containerEnd: Int) extends ParserState with HasExitAction {
+      @tailrec
+      final override def proceed(input: Byte, pp: JsonObjectParser): Unit = input match {
+        case _ if isWhitespace(input) ⇒
+          pp.take()
+          if (pp.keepSeeking) {
+            proceed(pp.nextInput, pp)
+          }
+
+        case _ if input == containerEnd ⇒
+          /* this is our end! */
+          pp.take()
+          exitAction(pp)
+
+        case DoubleQuote ⇒
+          pp.take()
+          pp.enterState(InString(this, pp))
+
+        case Comma | Colon ⇒
+          /* in a real JSON parser we'd check whether the colon and commas appear at appropriate places. Here
+          we do without: we're just framing JSON and this is good enough */
+          pp.take()
+          if (pp.keepSeeking) {
+            proceed(pp.nextInput, pp)
+          }
+
+        case CurlyBraceStart ⇒
+          pp.take()
+          pp.enterContainer(InObject, this)
+
+        case SquareBraceStart ⇒
+          pp.take()
+          pp.enterContainer(InArray, this)
+
+        case _ ⇒
+          pp.take()
+          pp.enterState(InNaked(this, pp))
+
+        //case _ =>
+        //  throw new FramingException(s"Invalid JSON encountered at position [${pp.pos}] of [${pp.buffer}]")
+      }
+    }
+
+    private object InArray extends InContainerBase(SquareBraceEnd) with LeaveContainerOnExit {
+      override def toString: String = "InArray"
```

Comment:
If you make these case objects they would get the toString "for free"

Reply:
Doesn't matter much but for the record I think you missed that I said **case** object - will be exactly like this, not a FQCN

Reply:
I did miss you said **case** object indeed! just pushed a correction.